### ISSUE TITLE
laggy zoom operation, especially with mouse wheel

### ISSUE THREAD

**Issue by o-smirnov** [Mulan-94/polarvis#6]
@ericmandel how is mouse wheel zoom working for you on the current version of the website? Mine seems very laggy.

Maybe it's not the mouse wheel per se, but rather that it takes about ~1.5 seconds for the image to re-render at the new zoom level. When using the toolbar zoom in/out buttons, this feels ok (because the zoom levels change by a factor of 2, and every zoom in or out is a discrete operation requiring two clicks from the user). But when using the mouse wheel, every notch of the wheel is a minor adjustment in zoom level, so if I move the wheel by, say, 4 clicks, it then takes 6 seconds for the display to catch up, as it patiently re-renders all the zooms in between.

**Comment by ericmandel** [comment]
@o-smirnov tbh, I forgot I had implemented support for mouse wheel until you opened this issue, I haven't used it in years and years, but I'll look into it as soon as I can extract myself from the bowels of socket.io. But tell me ... do you have 2K regions on the image when you're using the wheel? Cause now that you mention it, we'll need the same technique for zooming as we  used for panning, assuming that's possible.

It's an old algorithm and almost certainly can be improved. One thing you can try immediately is to change the value of
JS9.ADDZOOM (this should be in globalOpts, so this is VERY old). That's the zoom increment for each click of the wheel and is currently set to 0.1 ... try 0.25 or 0.5 and see if the response feels better.

**Comment by o-smirnov** [comment]
Yep, the 2000 regions are all being displayed. So it must be the same issue as in the panning.

I wouldn't just change the zoom increment though (been down this road with the Tigger viewer...) The real problem is that the wheel will issue a whole series of scroll events at the flick of a user's finger (often, an unintentionally long series). If the display doesn't respond to each event near-instantly, the visual cue to the user is broken. So if we up the zoom increment per event, but don't fix the lag, all we're doing is setting the user up for a series of unintentional super-zooms... done frustratingly slowly.

I would suggest one of two solutions. Both involve delaying the zoom:

* if zoom wheel events arrive in a quick series, ignore all but the last one. That is, don't do the zoom on the event itself, but start a timer of e.g. 0.5 secs... if the event repeats before the timeout is complete, ignore it.... and do the zoom when the timeout finally fires. This doesn't solve the lag, but it does prevent the user from "queing up" a series of zoom operations unintentionally.

* More fancy still: start the timer and draw a rectangular outline showing where you're going to zoom to. Update the outline with each subsequent wheel event -- so the user sees a shrinking box, indicating the "tentative" zoom region. Actually, the active region indicated on the panner should also shrink. Then when the wheel events stop coming (that timer again), execute the zoom.

  For zooming out, same thing, but expand the rectangular outline on the panner window to give the visual cue of the "tentative" unzoom.

Also, can we lock the zooming-out so that once the entire image is visible, it doesn't zoom out any further?

**Comment by ericmandel** [comment]
I'll be glad to implement one of the above as soon as I have finally and truly  fixed/unified the support for socket.io v2/v3 (right now we have both version inside the helper, but it looks like socket.io problems are fixed and pyjs9's python-socketio also is fixed and ...) And we'll deal with putting an upper limit on the zoom-out. It's great to get your suggestions on this, especially since I don't have much experience using it! Probably early next week ... keeping fingers crossed, of course.

**Comment by ericmandel** [comment]
OK, I lied ... implementing a delay in refreshing the shape layers during the "mouse wheel" action was straight-forward, so I did that this morning, and  created a test "mousewheel" branch in GitHub so you can check out the response (which looks pretty good to me.) Uses the same code as was used for the "panning the image" action. (I tried delaying the zoom itself but did not like it as much ... but I await your verdict.)

Be a little careful here: the "mousewheel" branch  contains the unified v2/v3 socket.io module layout. It really should be OK (and I expect to update into the main repository early next week), but for now, remember to update the node_modules directory, as well as js9Helper.js, if you're using the helper and don't care to do a full install.

I'll look into the zoom-out issue, maybe even today ...

**Comment by o-smirnov** [comment]
Cool, thanks @ericmandel! @Mulan-94, could you deploy please? And teach a man to fish, i.e. tell me how you've got it set up so I can try my own branches if I want to...

**Comment by ericmandel** [comment]
@o-smirnov zoom-out now stops once the entire image is visible (this is controlled by the globalOpts.mousetouchLimit property, which I believe can default to true.) It's updated in the mousewheel branch, so you can just "git pull', "git checkout mousewheel", etc.

**Comment by o-smirnov** [comment]
Thanks! I just don't want to mess with @Mulan-94's web page setup until I understand it. Lexy, I think we need to discuss and document this, there needs to be an easy way to pull updates.

**Comment by Mulan-94** [comment]
@o-smirnov yes, I agree. I have switched to the `mousewheel` branch and changed the js9.js, js9Helper.js and node_modules.

**Comment by Mulan-94** [comment]
@ericmandel @o-smirnov two questions: 1. Is there a way to free up a tab's memory incase a page with JS9 is reloaded multiple times? I notice that if I reload the same page multiple times with the same image, my RAM becomes very occupied at some point and the only way to free this memory is by closing the tab. I tried `JS9.CloseImage({clear: true});` but doesn't seem to cut it.

2. With each update in JS9, what do I need to re-install or copy? i.e how can I manage JS9 udpates?

**Comment by ericmandel** [comment]
> s there a way to free up a tab's memory incase a page with JS9 is reloaded multiple times? 

@Mulan-94 My only technique is: when the Chrome debugger is up, press and hold the reload key to see this menu and choose the hard reload:

<img width="461" alt="Screen Shot 2021-03-05 at 7 04 17 AM" src="https://user-images.githubusercontent.com/3028875/110113457-20d4c400-7d81-11eb-9721-fd775a794c3a.png">

Chrome has a button to trigger garbage collection but I can never find it ... in any case, I sometimes have to kill the whole Chrome program and start again, especially if I reload with breakpoints enabled.

**Comment by Mulan-94** [comment]
Oh Thank you, I didn't know about that. From what I see, the zoom seems much faster ...

**Comment by ericmandel** [comment]
> With each update in JS9, what do I need to re-install or copy? i.e how can I manage JS9 udpates?

Firstly: I'm only giving you branches because what we are doing has a subjective component ("is this good enough?") and I don't want to keep updating the main repository until you are more or less satisfied. The canonical process would be for me to update the main (master) repository and for you to do a 'git pull' to get the update.

In general, each time you update, I would recommend a script builds and installs in a location that you can then use for testing. On my Mac development machine, one of my configurations is that I install in ~/Sites and load pages from there. This should only take a minute ...

(But for everyday development, I just start chrome with the --args --allow-file-access-from-files switch and use the file:// URI in the directory which I just pulled into. This works if you a clean separation between JS9 and your own code, i.e. your code is in a separate directory and you load a web page that loads JS9 resources relative to your own resources. Not sure I recommend this, but it is quick and easy!)

**Comment by ericmandel** [comment]
> Thanks! I just don't want to mess with @Mulan-94's web page setup until I understand it. 

All of that said, I could update the main repository today if time is pressing.  I generally don't like to do that on Fridays, as it can make for a very bad weekend, but I'm open to your schedule ...

**Comment by Mulan-94** [comment]
> In general, each time you update, I would recommend a script builds and installs in a location that you can then use for testing. 

Noted.
> All of that said, I could update the main repository today if time is pressing. I generally don't like to do that on Fridays, as it can make for a very bad weekend, but I'm open to your schedule ...

I'll let @o-smirnov give his opinion on this one first ...

**Comment by ericmandel** [comment]
@Mulan-94 This loaded in about 10 seconds on my super-slow DSL line, very nice! I'll update the main repository on Github early next week and let you know, in case you want to install from the main branch (the code will be the same!)

(p.s. I still have my much-faster-if-only-it-would-actually-work refresh code and will keep thinking about it.)

Here are two aesthetic nits in your really nice web page, as seen from my 15" laptop:

I'm a little puzzled about why you are getting this (second line of the toolbar? or an old toolbar config?):

<img width="624" alt="Screen Shot 2021-03-05 at 1 39 10 PM" src="https://user-images.githubusercontent.com/3028875/110159513-ceaf9500-7db8-11eb-87c3-11d12256ce3e.png">

instead of this:

<img width="552" alt="Screen Shot 2021-03-05 at 1 50 28 PM" src="https://user-images.githubusercontent.com/3028875/110160392-f94e1d80-7db9-11eb-9179-05cf777aea55.png">

2. There is some overlap between plugins and captions, you might want to change the size or layout or something, or just remove the obscured captions completely?
 
<img width="621" alt="Screen Shot 2021-03-05 at 1 39 19 PM" src="https://user-images.githubusercontent.com/3028875/110160011-7036e680-7db9-11eb-9d37-eabbfb2f639f.png">

**Comment by o-smirnov** [comment]
Yep, and the zoom is a lot snappier. I still tend to "overzoom" if I'm not careful with the mouse wheel, but if I take it slow, there's enough of a visual feedback loop.

@ericmandel, I was wondering: do you get x,y positions along with the mouse wheel event? If so, it would be nice to zoom into the position that the mouse is at, rather than zooming into the current image centre.

**Comment by ericmandel** [comment]
Oh wait, the toolbar is simply too large in height, so you are seeing two lines of toolbar, instead of one. That should be easily fixable ...

**Comment by ericmandel** [comment]
> I still tend to "overzoom" if I'm not careful with the mouse wheel, 

For some reason, on this page, the zoom just gets out of hand on my Apple mouse. This doesn't happen on my test web page, the mouse-based scrolling is much better behaved ... not sure what the difference is ...

> it would be nice to zoom into the position that the mouse is at, rather than zooming into the current image centre.

Assuming the position info is available, this would seem to require a pan as well as a zoom ... would it end up panning with every mouse-wheel click, or could it be smarter? I'll have a look at the possibilities ...

**Comment by ericmandel** [comment]
> I still tend to "overzoom" if I'm not careful with the mouse wheel,

OK I found a test file that displays overzoom, so I'll what adding your original timeout fix does ...

**Comment by ericmandel** [comment]
Hmmm ... I'm appending a js9.js (renamed to js9.txt to make GitHub happy) that has the delayed wheel zoom. It's not subject to pileup, but seems to be pretty slow/unresponsive to my mouse "wheel", even when I shortened the timeout to make it run more often. Not sure what to do here ...


[js9.txt](https://github.com/Mulan-94/polarvis/files/6093035/js9.txt)

**Comment by o-smirnov** [comment]
> Assuming the position info is available, this would seem to require a pan as well as a zoom ... would it end up panning with every mouse-wheel click, or could it be smarter?

I was thinking pan to the mouse position, and zoom in, but in a single operation. It's still just one re-rendering, so it shouldn't be any slower, right? 

But speaking of smarter, one thing to be careful about: say I click the wheel at a physical x,y position, and the image pans and zooms in. My mouse is now at the same spot on the screen, but at a different WCS location on the image relative to where I initiated the zoom. So if I click the wheel again, I'm now panning to a different spot on the image. Hmmm, that might not work so well. Perhaps the clever thing to do is a pan-and-zoom that ensures that the same WCS position remains anchored to the screen position of the mouse? Am I making sense?

**Comment by ericmandel** [comment]
Are you aware that you can shift-click to pan to the mouse position? Perhaps using that in conjunction with the mouse wheel will suffice? I think it keeps the center constant as it zooms.

**Comment by ericmandel** [comment]
(I'm not enamored of shifting in the same operation as zoom, it does seem like its going to get confusing ...)

**Comment by o-smirnov** [comment]
Ah no, I didn't know about shift-click. I don't quite understand what it does though, I just get two lines like so:

![image](https://user-images.githubusercontent.com/6470079/110169206-0cd0a700-7e01-11eb-9c67-4d43b2b0613e.png)

...but it doesn't pan. Also, if I shift-click on a region, it selects the region just like a normal click, which is a bit confusing.

**Comment by o-smirnov** [comment]
Ah wait, but Ctrl+click pans...

**Comment by ericmandel** [comment]
Oh sorry! My ctrl key is too stiff for my fingers, so I remapped shift to be crtl ... yes, its ctrl-click!

btw, you are seeing the cross-hair cursor, which you get with (after  toggling the 'c' key) when pressing shift and moving the mouse.

**Comment by o-smirnov** [comment]
Ah ok I see. I'm just discovering the cross-hair -- what else can I do with it?

**Comment by o-smirnov** [comment]
As I look at all this, I really think it needs to be a paper in Astronomy & Computing or something... @Mulan-94, feel like writing something up?

**Comment by ericmandel** [comment]
You can wcs match crosshairs between two related images using the View -> show -> match wcs crosshairs option. Turn on the cross hair on both images, select match wcs crosshairs, and when you move the crosshair in one image, the other one moves as well.

**Comment by ericmandel** [comment]
> As I look at all this, I really think it needs to be a paper in Astronomy & Computing or something... @Mulan-94, feel like writing something up?

You know what would be really fantastic: if @Mulan-94 would consider, in consultation with @o-smirnov, taking some initial steps toward joining me in developing JS9. This past year has really brought home the risks of one-person projects and it might be good to bring on a trust-worthy person ...

**Comment by o-smirnov** [comment]
Oh, I'm all for that. I'm too old to learn Javascript quickly, so I just keep poking around the edges, but @Mulan-94 is the right age for it. ;)

**Comment by ericmandel** [comment]
And the right brains for it (admittedly, as seen from a distance of 8125 miles)

**Comment by ericmandel** [comment]
Ah ... I added a mod factor (currently 2) to skip mousewheel clicks, and the result is less pileup and smoother zooming. I updated this into the mousewheel branch, let me know what you think on Monday ...

**Comment by o-smirnov** [comment]
Smoother indeed, I like it!

One thing I keep meaning to ask -- it feels like the direction of the zoom is inverted, do we have an option to swap that around? 

I was trying to think why it felt inverted, then I realized it -- in the browser itself (at least in Chrome), I can click Ctrl and use the wheel to "zoom" into websites, but I scroll **up** for zoom **in** and down for zoom out. So I think that's the more common convention.

**Comment by ericmandel** [comment]
I parameterized the direction and changed the default ... you might recall that browsers changed the zoom and scroll direction 20+ years ago, and I've been confused ever since (Because I was there at the birth of X11, and the X defaults, which are opposite, were forever imprinted??)

Its updated in the mousehweel branch ...

**Comment by o-smirnov** [comment]
Ah yes, I do have vague memories of some confusion about it. But surely the mouse wheel came a lot later than X11? I distinctly remember a time when a three-button mouse (as opposed to 2) was the height of sophistication, and X11 made good use of those buttons.

**Comment by ericmandel** [comment]
It actually had to do the direction of scrolling when you moved the scroll-bar widget ... the mouse wheel just followed on in the same way ...

Fun fact: the original saoimage circa late 1980's (followed by saotng, ds9, and now js9, still trying to get it right!) was the first program written  using X outside the MIT Athena project. One of the X architects, Jim Gettys, worked here before moving to DEC and the Athena project ... those were pretty fun and heady days ...

**Comment by ericmandel** [comment]
FYI, the main JS9 repository now has all the code from the mousewheel branch (plus a bunch of other housekeeping-type updates to underlying packages). I'll remove the mousewheel branch pretty soon ... while trying to get up the courage to release v3.4 ...
